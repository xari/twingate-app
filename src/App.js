import { useState, useCallback } from "react";
import { classNames, isJsonString, prettyPrint } from "./utils";
import Hero from "./Hero";
import ImgTxt from "./ImgTxt";
import Data from "./Data";
import "./App.css";

function App() {
  const [error, setError] = useState(false);
  const [JSONStr, setJSONStr] = useState(
    prettyPrint(
      JSON.stringify([
        {
          type: "hero",
          imageURI:
            "https://images.unsplash.com/photo-1579963333765-b4129b3250fc",
        },
        {
          type: "image-text",
          imageURI:
            "https://images.unsplash.com/photo-1579963333765-b4129b3250fc",
          text: "Sunrise from the ground......",
          title: "Beach",
          leftToRight: true,
        },
        {
          type: "image-text",
          imageURI:
            "https://images.unsplash.com/photo-1579963333765-b4129b3250fc",
          text: "Sunset from the sky......",
          title: "Airplane",
          leftToRight: false,
        },
        {
          type: "data",
          url: "https://api.publicapis.org/random",
        },
      ])
    )
  );

  const handleBlur = useCallback(
    (event) => {
      const value = event.target.value;

      if (isJsonString(value)) {
        setError(false);

        // If there's a change to the value, then update state
        value !== JSONStr && setJSONStr(prettyPrint(value));
      } else {
        setError(true);
      }
    },
    [JSONStr]
  );

  return (
    <div className="grid grid-cols-2">
      <div className="">
        {error && (
          <span>It looks like something is not right with your JSON.</span>
        )}
        <textarea
          className={classNames(
            error === false ? "bg-neutral-50" : "bg-red-50",
            "h-screen w-full block"
          )}
          cols={50}
          rows={30}
          defaultValue={JSONStr}
          onBlur={handleBlur}
          style={{ resize: "none" }}
        ></textarea>
      </div>
      <div className="">
        {JSON.parse(JSONStr).map((props, i) => {
          switch (props.type) {
            case "hero":
              return <Hero key={i} {...props} />;
              break;
            case "image-text":
              return <ImgTxt key={i} {...props} />;
              break;
            case "data":
              return <Data key={i} {...props} />;
              break;
            default:
              return <span>Non standard type</span>;
          }
        })}
      </div>
    </div>
  );
}

export default App;
