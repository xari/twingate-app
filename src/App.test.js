import renderer from "react-test-renderer";
import App from "./App";

test("renders the App", () => {
  const tree = renderer.create(<App />).toJSON();

  expect(tree).toMatchSnapshot();
});
