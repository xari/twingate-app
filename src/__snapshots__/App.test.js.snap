// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the App 1`] = `
<div
  className="grid grid-cols-2"
>
  <div
    className=""
  >
    <textarea
      className="bg-neutral-50 h-screen w-full block"
      cols={50}
      defaultValue="[
    {
        \\"type\\": \\"hero\\",
        \\"imageURI\\": \\"https://images.unsplash.com/photo-1579963333765-b4129b3250fc\\"
    },
    {
        \\"type\\": \\"image-text\\",
        \\"imageURI\\": \\"https://images.unsplash.com/photo-1579963333765-b4129b3250fc\\",
        \\"text\\": \\"Sunrise from the ground......\\",
        \\"title\\": \\"Beach\\",
        \\"leftToRight\\": true
    },
    {
        \\"type\\": \\"image-text\\",
        \\"imageURI\\": \\"https://images.unsplash.com/photo-1579963333765-b4129b3250fc\\",
        \\"text\\": \\"Sunset from the sky......\\",
        \\"title\\": \\"Airplane\\",
        \\"leftToRight\\": false
    },
    {
        \\"type\\": \\"data\\",
        \\"url\\": \\"https://api.publicapis.org/random\\"
    }
]"
      onBlur={[Function]}
      rows={30}
      style={
        Object {
          "resize": "none",
        }
      }
    />
  </div>
  <div
    className=""
  >
    <div>
      <img
        className="object-cover h-48 w-full"
        src="https://images.unsplash.com/photo-1579963333765-b4129b3250fc"
      />
    </div>
    <div
      className="grid grid-cols-2 w-full"
    >
      <div
        className="h-full w-full bg-cover bg-center"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1579963333765-b4129b3250fc)",
          }
        }
        title="Beach"
      />
      <div
        className="h-48 p-5"
      >
        <div
          className="font-bold text-xl mb-2"
        >
          Beach
        </div>
        <p>
          Sunrise from the ground......
        </p>
      </div>
    </div>
    <div
      className="grid grid-cols-2 w-full text-right"
    >
      <div
        className="h-48 p-5"
      >
        <div
          className="font-bold text-xl mb-2"
        >
          Airplane
        </div>
        <p>
          Sunset from the sky......
        </p>
      </div>
      <div
        className="h-full w-full bg-cover bg-center"
        style={
          Object {
            "backgroundImage": "url(https://images.unsplash.com/photo-1579963333765-b4129b3250fc)",
          }
        }
        title="Airplane"
      />
    </div>
    <div
      className="grid grid-cols-5 gap-4 h-48 p-1 overflow-hidden"
    >
      <div
        className="col-span-1"
      >
        <button
          className="items-center px-2.5 py-1.5 border border-transparent text-xs font-medium rounded shadow-sm text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
          onClick={[Function]}
          type="button"
        >
          Refresh
        </button>
      </div>
      <pre
        className=""
      >
        Loading...
      </pre>
    </div>
  </div>
</div>
`;
